### PR TITLE
Initial unity project and base functionality

### PR DESCRIPTION
**Pull request by Simoyd**
has string[] property to populate the object.
currently fills entire canvas (needs rect transform in future).
colors not currently being parsed, but commented code stubbed for how to update colors efficiently.
resizing logic not implemented.

### CODE REVIEW

**Comment by seanmakesgames** [comment]
lgtm (looks good to me)
code review process will adapt and change over time, but for now I'll get another set of eyes to sign off on it.

**Comment by Cheshyr** [comment]
None of the following suggestions are worth rejecting the PR.  They are only trivial tweaks and optimizations.

I would prefer more parenthesis to make logic evaluations explicit.  I know most people are ok with implicit operand order, but beung explicit where you can seems safer.

e.g. scroll_test.cs, lines 78 and 86.

if (engine.CorruptedLines == 0 || engine.CorruptedLines < lastCorrCount)
vs
if ((engine.CorruptedLines == 0) || (engine.CorruptedLines < lastCorrCount))

OnGUI can be called several times PER FRAME, so we may want to cache variable declarations in the parent class, as opposed to in the function call, to reduce burden on the stack.  It won't matter on this small scale test, but it's a simple optimization that prevents it from becoming a potential problem later.  If I'm mistaken on how this work, apologies.

e.g. fps_display.cs, lines 17, 19, 21, 25... etc.

public class fps_display : MonoBehaviour
{
	float deltaTime = 0.0f;
	int w;
	GUIStyle style;
	Rect rect;
	float msec;
	float fps;
	string text;
	
	...
	
	void OnGUI()
	{
		w = Screen.width, h = Screen.height;
		style = new GUIStyle();
		rect = new Rect(0, 0, w, h * 2 / 25);
		
		...
		
		
Technically, this could also be applied to elapsed, orgMode, and msPerLine in scroll_test.cs.

**Comment by ethankaminski** [comment]
`Assets/testing/scripts/fps_display.cs` and `Assets/testing/scripts/scroll_test.cs` both look reasonable.

Would it be possible to remove the `using System.Diagnostics;` line from `Assets/u3d_text_3ngine/scripts/u3d_text_3ngine.cs`? Or pull out the individual things that you use as individual `using X = ...` statements? That file looks large and in some places a bit dense, and I recall that `System.Diagnostics` provides access to stuff like running external programs. I'd feel more comfortable if it were easier to verify that the dangerous stuff isn't being used.

**Comment by ethankaminski** [comment]
Also, since Cheshyr mentioned style stuff: kudos for using explicit blocks in single-line control structure bodies. I find that clearer and more foolproof than omitting the block. :+1:

**Comment by Cheshyr** [comment]
u3d_text_3ngine.cs has similar cached variable opportunities.  For instance, CheckRecalibrate() creates a new canvas, templateRect, and templateGui on every frame (lines 236, 238, 242).  There are a few other instances later in the file.  Not worries about simple variables, but complex objects can get expensive, so we may want to optimize those early, and get them into the next commit.

**Comment by Cheshyr** [comment]
to confirm, LGTM.

**Comment by Simoyd** [comment]
CheckRecalibrate is only run if the cotnrol is resized. The boolean checked at the top of the function is set to false and never ever set back to true because I'm not detecting resizing yet.

**Comment by Simoyd** [comment]
Fixed the sys.diag stuff @ethankaminski

**Comment by Simoyd** [comment]
Cleaned up fps_display as per @Cheshyr comments. more "new <object>" reductions can happen in other files as we go.

**Comment by seanmakesgames** [comment]
LGTM

File: Assets/u3d_text_3ngine/scripts/u3d_text_3ngine.cs [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+    private List<GameObject> unusedLinesAge = new List<GameObject>();
+
+    /// <summary>
+    /// The TMP objects not currently being displayed on the screen (other than corruption lines)
+    /// </summary>
+    /// <remarks>Stored by original string they represent</remarks>
+    private Dictionary<string, List<GameObject>> unusedLinesLookup = new Dictionary<string, List<GameObject>>();
+
+    /// <summary>
+    /// The corruption lines that are not currently ont he screen
```
Comment (n00bish) [inline Assets/u3d_text_3ngine/scripts/u3d_text_3ngine.cs]: typo

**Comment by ethankaminski** [comment]
LGTM now that the System.Diagnostics change went through.

**Comment by ethankaminski** [comment]
It seems like Github's web view has a width of about 128 characters, by the way; might be worth hard-wrapping to that length in future?

**Comment by seanmakesgames** [comment]
Agreed. Weigh in on the code review stuff #8 and we can start hashing out the code quality rules for the repo.

**Comment by seanmakesgames** [comment]
Any other changes you want in this PR before I merge it? @Simoyd

**Comment by Simoyd** [comment]
No. I think we should get this in. I will be making a separate pull request soon with pri1 and pri2 done. @seanmakesgames

**Pull request merged**